### PR TITLE
Legacy csvV2 calibrations for 2018 PbPb data

### PR DESCRIPTION
**Pull request by mandrenguyen**
#### PR description:

This PR adds csvV2 b-tagging calibrations for the 2018 PbPb data.  The calibrations are delivered via an xml file and require a ESProducer to read them.  This xml file is included in 
https://github.com/cms-data/RecoBTag-Combined/pull/36
  For Run 3, we plan to resynchronize with the b-tagging calibrations used for pp.   However, these legacy calibrations  are needed for the 2018 PbPb miniAOD to match what is currently being used at analysis level.  

#### PR validation:

Tested on 140.5611

### CODE REVIEW

**Comment by slava77** [comment]
> The calibrations are delivered via an xml file and require a ESProducer to read them.

the file  RecoBTag/Combined/data/TMVA_Btag_CsJets_PbPb2018_BDTG.weights.xml should be removed from this PR and proposed as a part of a PR in https://github.com/cms-data/RecoBTag-Combined
Please scrub the xml file from the commit history (given one commit, a rebase will be needed).

**Comment by mandrenguyen** [comment]
@slava77 Sorry for putting the xml file in the wrong place.
How do I properly remove it from the commit history?  
I don't want to mess that up.

File: RecoBTag/Combined/interface/hiRun2LegacyCSVv2Tagger.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+#include <memory>
+#include <type_traits>
+
+/** \class hiRun2LegacyCSVv2Tagger
+ *  \author M. Nguyen
+ *  copied from CharmTagger.h (by M. Verzetti)
+ */
+
+class hiRun2LegacyCSVv2Tagger : public JetTagComputer {
```
Comment (slava77) [inline RecoBTag/Combined/interface/hiRun2LegacyCSVv2Tagger.h:20]: - class name should be capitalized
- what makes it Run2?
- what makes it Legacy?
- what makes it v2? 

I'd call it HeavyIonsCSVTagger or HeavyIonsCSVv1Tagger if we want to commit to versioning for simultaneous support of multiple versions.

**Comment by slava77** [comment]
> @slava77 Sorry for putting the xml file in the wrong place.
> How do I properly remove it from the commit history?
> I don't want to mess that up.

smth like 
```
git reset --soft HEAD~2 #resets to before all your 2 commits (as in the PR now)
git reset HEAD RecoBTag/Combined/data/TMVA_Btag_CsJets_PbPb2018_BDTG.weights.xml #undo git add
rm RecoBTag/Combined/data/TMVA_Btag_CsJets_PbPb2018_BDTG.weights.xml
git commit 
scram build code-checks; scram build code-format
git push --force your-remote-name hiLegacyRun2CSVv2
```

you may want to also modify the file names given https://github.com/cms-sw/cmssw/pull/31660/files#r499205087

File: RecoBTag/Combined/interface/hiRun2LegacyCSVv2Tagger.h (hunk 1 above)
Comment (mandrenguyen) [reply RecoBTag/Combined/interface/hiRun2LegacyCSVv2Tagger.h:20]: This is the csvV2 tagger, as defined by CMS for the outside world.   The version signifies both the training variables and the training method.  
This version of the tagger was used for the 2015 and 2018 PbPb data. 
It's now 'legacy' b/c it's based on quite outdated methods, which are not really supported in pp anymore.  It's legacy in the sense that after this reMiniAOD of the 2018 PbPb data, we will not support this for Run 3, or for any further processing of old data.   From everything we've seen so far, there should not be any need for dedicated heavy-ion b-tagging code, so I don't foresee future versions of this.

Comment (slava77) [reply RecoBTag/Combined/interface/hiRun2LegacyCSVv2Tagger.h:20]: I suppose we'll need to iterate. I'd rather have similar naming pattern to the existing one.
Also, the pp CSV is apparently in RecoBTag/SecondaryVertex.

If this tagger was used on 2015 data, is hiRun2LegacyCSVv2Tagger something that was named this way in analysis recipes?

For the old pp track-based CSV tagger
- apparently there is no V2 C++ class, just the configuration (assuming this refers to CombinedSVComputer)
- is "Legacy" really meant to reflect that this is using tracks, not candidates? Then the non-HI version comes without the extra modifiers and we better stay closer to what's available.

Comment (mandrenguyen) [reply RecoBTag/Combined/interface/hiRun2LegacyCSVv2Tagger.h:20]: Yeah, I realized after my first attempt that the pp one is under SecondaryVertex.  
But when I moved the code there I ran into an issue related to the declaration of CombinedSVSoftLeptonComputer sl_computer_;
The variable SoftLeptonFlip gets defined twice or something. 
I wasn't able to easily figure it out. 

CSVv2 is the name of the tagger, even for the world outside CMS. 
The Run2Legacy bit provides extra info, but if you don't like, how about we go for   "HeavyIonCSVv2" ?

**Comment by slava77** [comment]
I think this better be reviewed by the BTV experts
The goal would be to maximize reuse of existing code and to also have a similar implementation and naming patterns as in the existing CSVv2.
So that someone familiar with the pp CSVv2 (even the track-based one) would be able to see the relationship.

@emilbols , please take a look or pass it to the experts who worked/developed CSVv2 in the past.

**Comment by emilbols** [comment]
@mandrenguyen Could you prepare a presentation on the technical implementation of this PR at the next BTV software meeting (wednesday the 14th)? Mainly so we can better understand what the bottlenecks are for reusing the producer for CSVv2, since it seems you are switching to a version of the producer used for the charm tagger. Also CSVv2 consists of three trainings on samples of jets with NoVertex, a PseudoVertex and a RecoVertex, and these are then combined. As far as i can see, you do not do that here (which is probably better, the old method is outdated). However it would be good, if you can highlight how the training is exactly done (and what ML framework you use), so we can see what the overlap with CSVv2 is.

**Comment by mandrenguyen** [comment]
@emilbols   
The training was done with a BDT, and yes, we did not use the 3 categories. 
The training we first done by Stanislav Lisniak for the 2015 PbPb data and then by Inna Kucher for the 2018 PbPb data.   They both presented in BTV and are both are now out of CMS.
I'm afraid I don't have any further details beyond what was last presented in BTV:
https://indico.cern.ch/event/853820/contributions/3590366/attachments/1921458/3178704/HIN_BTag2018_07102019.pdf
I'm simply trying to port what we currently use in the heavy-ion analysis recipes into miniAOD.

File: RecoBTag/Combined/python/hiRun2LegacyCSVv2_helpers.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import os
+from RecoBTag.Combined.hiRun2LegacyCSVv2_trainingvars import get_var_pset
+import xml.etree.ElementTree as ET
+from pdb import set_trace
```
Comment (jpata) [inline RecoBTag/Combined/python/hiRun2LegacyCSVv2_helpers.py:4]: remove

File: RecoBTag/Combined/python/hiRun2LegacyCSVv2_helpers.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+def get_path(file_in_path):
```
Comment (jpata) [inline RecoBTag/Combined/python/hiRun2LegacyCSVv2_helpers.py:6]: why not import them from RecoBTag/CTagging/python/helpers.py?

File: RecoBTag/Combined/src/hiRun2LegacyCSVv2Tagger.cc [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+#include "DataFormats/BTauReco/interface/CandIPTagInfo.h"
+#include "DataFormats/BTauReco/interface/SecondaryVertexTagInfo.h"
+#include "FWCore/Utilities/interface/ESInputTag.h"
+#include "CondFormats/DataRecord/interface/GBRWrapperRcd.h"
+
+#include <iostream>
+#include <memory>
+
+#include <algorithm>
+#include <iostream>
```
Comment (jpata) [inline RecoBTag/Combined/src/hiRun2LegacyCSVv2Tagger.cc:13]: remove both iostream imports, should not be needed after removing couts.

**Review by jpata — COMMENTED** [review]
some initial, mostly technical comments inline.

File: RecoBTag/Combined/python/hiRun2LegacyCSVv2Computer_cfi.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import FWCore.ParameterSet.Config as cms
+import RecoBTag.SecondaryVertex.candidateCombinedSecondaryVertexV2Computer_cfi as sl_cfg 
+from RecoBTag.Combined.hiRun2LegacyCSVv2_trainingSettings import hiRun2LegacyCSVv2_vpset
+from RecoBTag.Combined.hiRun2LegacyCSVv2_helpers import get_vars
```
Comment (emilbols) [inline RecoBTag/Combined/python/hiRun2LegacyCSVv2Computer_cfi.py:4]: Why do you need this get_vars function? It looks like it is not used for anything, so please remove the import. And as far as i can tell you copied two scripts from the c-tagger, i.e. the hiRun2LegacyCSVv2_trainingvars and hiRun2LegacyCSVv2_helpers files just to get this function. Can we not remove those scripts? If you need them for tests you can just use the ones in RecoBTag.CTagging

File: RecoBTag/Combined/python/hiRun2LegacyCSVv2Computer_cfi.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import FWCore.ParameterSet.Config as cms
+import RecoBTag.SecondaryVertex.candidateCombinedSecondaryVertexV2Computer_cfi as sl_cfg
```
Comment (emilbols) [inline RecoBTag/Combined/python/hiRun2LegacyCSVv2Computer_cfi.py:2]: can you rename sl_cfg to sv_cfg. The sl stands for soft lepton which is used in the c-tagger, but here you are just using the SV+track info, so the consistent naming with btv code would be sv

File: RecoBTag/Combined/python/hiRun2LegacyCSVv2Computer_cfi.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+import RecoBTag.SecondaryVertex.candidateCombinedSecondaryVertexV2Computer_cfi as sl_cfg 
+from RecoBTag.Combined.hiRun2LegacyCSVv2_trainingSettings import hiRun2LegacyCSVv2_vpset
+from RecoBTag.Combined.hiRun2LegacyCSVv2_helpers import get_vars
+
+
+hiRun2LegacyCSVv2Computer = cms.ESProducer(
+   'hiRun2LegacyCSVv2ESProducer',
+   slComputerCfg = cms.PSet(
+      **sl_cfg.candidateCombinedSecondaryVertexV2Computer.parameters_()
+      ),
```
Comment (emilbols) [inline RecoBTag/Combined/python/hiRun2LegacyCSVv2Computer_cfi.py:11]: change to svComputerCfg and sv_cfg

File: RecoBTag/Combined/python/hiRun2LegacyCSVv2Computer_cfi.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+   weightFile = cms.FileInPath('RecoBTag/Combined/data/TMVA_Btag_CsJets_PbPb2018_BDTG.weights.xml'),
+
+   variables = hiRun2LegacyCSVv2_vpset,
+   computer = cms.ESInputTag('dummy:dummy'),
```
Comment (emilbols) [inline RecoBTag/Combined/python/hiRun2LegacyCSVv2Computer_cfi.py:15]: is this used? i think you can just remove this computer line

File: RecoBTag/Combined/src/hiRun2LegacyCSVv2Tagger.cc [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    gbrForest_ = cc.consumes(edm::ESInputTag{"",
+                                             configuration.existsAs<std::string>("gbrForestLabel")
+                                                 ? configuration.getParameter<std::string>("gbrForestLabel")
+                                                 : ""});
+  }
+}
+
+hiRun2LegacyCSVv2Tagger::hiRun2LegacyCSVv2Tagger(const edm::ParameterSet &configuration, Tokens tokens)
+    : sl_computer_(configuration.getParameter<edm::ParameterSet>("slComputerCfg")),
+      sv_computer_(configuration.getParameter<edm::ParameterSet>("slComputerCfg")),
```
Comment (emilbols) [inline RecoBTag/Combined/src/hiRun2LegacyCSVv2Tagger.cc:30]: remove the sl_computer_ and update the config name
```suggestion
    : sv_computer_(configuration.getParameter<edm::ParameterSet>("svComputerCfg")),
```

File: RecoBTag/Combined/interface/hiRun2LegacyCSVv2Tagger.h [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    std::string name;
+    reco::btau::TaggingVariableName id;
+    size_t index;
+    bool has_index;
+    float default_value;
+  };
+
+private:
+  std::unique_ptr<TMVAEvaluator> mvaID_;
+  CombinedSVSoftLeptonComputer sl_computer_;
```
Comment (emilbols) [inline RecoBTag/Combined/interface/hiRun2LegacyCSVv2Tagger.h:45]: Should not be needed, remove it.

File: RecoBTag/Combined/interface/hiRun2LegacyCSVv2Tagger.h [hunk 11]
```diff
@@ -0,0 +1,56 @@
+#ifndef RecoBTag_Combined_hiRun2LegacyCSVv2Tagger_h
+#define RecoBTag_Combined_hiRun2LegacyCSVv2Tagger_h
+
+#include "FWCore/Framework/interface/ESConsumesCollector.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "CommonTools/MVAUtils/interface/TMVAEvaluator.h"
+#include "RecoBTau/JetTagComputer/interface/JetTagComputer.h"
+#include "RecoBTag/SecondaryVertex/interface/CombinedSVSoftLeptonComputer.h"
```
Comment (emilbols) [inline RecoBTag/Combined/interface/hiRun2LegacyCSVv2Tagger.h:8]: You should be able to remove this include

**Review by emilbols — COMMENTED** [review]
Given this model uses a different training setup, framework and model format from the old CSV trainings, i think it makes sense to setup a new producer, it would not be that straightforward to try and fit this into the mostly deprecated setup used for pp CSV. However i think we need a bit of cleanup, since some of this code copied from the c-tagger implementation, does not seem necessary here.

For the naming i would say the most consistent with past modules seems to be to use something like HeavyIonsCSVTagger because while this model is very similar to CSVv2, it is not identical, given that it is a bdt trained on the inclusive sample compared to CSVv2 where you use a shallow nn to combine vertex categories. This exact model architecture is not used in BTV anywhere other than for Heavy Ions, so i think naming it HeavyIonsCSV to indicate it is a CSV variant for heavy ions makes the most sense. We do not have something like Run2Legacy in the class name of any of our other producers, so i would agree we remove this as well.

File: RecoBTag/Combined/interface/heavyIonCSVTagger.h [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    std::string name;
+    reco::btau::TaggingVariableName id;
+    size_t index;
+    bool has_index;
+    float default_value;
+  };
+
+private:
+  std::unique_ptr<TMVAEvaluator> mvaID_;
+  //CombinedSVSoftLeptonComputer sl_computer_;
```
Comment (jpata) [inline RecoBTag/Combined/interface/heavyIonCSVTagger.h:46]: remove commented-out code

File: RecoBTag/Combined/src/heavyIonCSVTagger.cc [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+HeavyIonCSVTagger::Tokens::Tokens(const edm::ParameterSet &configuration, edm::ESConsumesCollector &&cc) {
+  if (configuration.getParameter<bool>("useCondDB")) {
+    gbrForest_ = cc.consumes(edm::ESInputTag{"",
+                                             configuration.existsAs<std::string>("gbrForestLabel")
+                                                 ? configuration.getParameter<std::string>("gbrForestLabel")
+                                                 : ""});
+  }
+}
+//sl_computer_(configuration.getParameter<edm::ParameterSet>("slComputerCfg")),
```
Comment (jpata) [inline RecoBTag/Combined/src/heavyIonCSVTagger.cc:20]: remove

File: RecoBTag/Combined/src/heavyIonCSVTagger.cc [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+  bool notTaggable = false;
+  bool noTrack = false;
+  float vtxMassVal = 0.;
+
+  for (auto &mva_var : variables_) {
+    //vectorial tagging variable
+    if (mva_var.has_index) {
+      std::vector<float> vals = vars.getList(mva_var.id, false);
+      inputs[mva_var.name] = (vals.size() > mva_var.index) ? vals[mva_var.index] : mva_var.default_value;
+      if (mva_var.name == "TagVarCSV_trackSip3dSig_0" && inputs[mva_var.name] < -98.999) noTrack = true;
```
Comment (jpata) [inline RecoBTag/Combined/src/heavyIonCSVTagger.cc:91]: as suggested previously, either remove the string comparison that is done at every evaluation, or check with igprof that strcmp does not show up in the profiling. Please add a comment for the origin of the magic number -98.99 at least.

File: RecoBTag/Combined/src/heavyIonCSVTagger.cc [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    if (mva_var.has_index) {
+      std::vector<float> vals = vars.getList(mva_var.id, false);
+      inputs[mva_var.name] = (vals.size() > mva_var.index) ? vals[mva_var.index] : mva_var.default_value;
+      if (mva_var.name == "TagVarCSV_trackSip3dSig_0" && inputs[mva_var.name] < -98.999) noTrack = true;      
+    }
+    //single value tagging var
+    else {
+      inputs[mva_var.name] = vars.get(mva_var.id, mva_var.default_value);
+      //IK: vtxMass check to check vtxType: vtxType = 2 (no vtx), vtxMass < 0, vtxType = 1 (pseudo vtx), vtxMass > 0
+      if (mva_var.name == "TagVarCSV_vertexMass") vtxMassVal = inputs[mva_var.name];
```
Comment (jpata) [inline RecoBTag/Combined/src/heavyIonCSVTagger.cc:97]: same as above for string comparison

File: RecoBTag/Combined/interface/heavyIonCSVTagger.h [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+class HeavyIonCSVTagger : public JetTagComputer {
+public:
+  struct Tokens {
+    Tokens(const edm::ParameterSet& configuration, edm::ESConsumesCollector&& cc);
+    edm::ESGetToken<GBRForest, GBRWrapperRcd> gbrForest_;
+  };
+
+  /// explicit ctor
+  HeavyIonCSVTagger(const edm::ParameterSet&, Tokens);
+  ~HeavyIonCSVTagger() override;  //{}
```
Comment (jpata) [inline RecoBTag/Combined/interface/heavyIonCSVTagger.h:30]: please confirm if/why this is needed

File: RecoBTag/Combined/interface/heavyIonCSVTagger.h [hunk 17]
```diff
@@ -0,0 +1,57 @@
+#ifndef RecoBTag_Combined_heavyIonCSVTagger_h
+#define RecoBTag_Combined_heavyIonCSVTagger_h
+
+#include "FWCore/Framework/interface/ESConsumesCollector.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "CommonTools/MVAUtils/interface/TMVAEvaluator.h"
+#include "RecoBTau/JetTagComputer/interface/JetTagComputer.h"
+#include "RecoBTag/SecondaryVertex/interface/CombinedSVComputer.h"
+//#include "RecoBTag/SecondaryVertex/interface/CombinedSVSoftLeptonComputer.h"
```
Comment (jpata) [inline RecoBTag/Combined/interface/heavyIonCSVTagger.h:9]: remove

**Review by jpata — COMMENTED** [review]
Thanks for the cleanup suggested by @emilbols above, looks like good progress. I had a few more comments, some of them are repeats of the above that were not resolved so far.

Minor, but why heavyIonCSVTagger.cc/h and not HeavyIonCSVTagger.cc/h to be consistent with the class name and other files in the directory?

Can you also please update the PR description with a TimeMemoryInfo or igprof output?

File: RecoBTag/Combined/src/heavyIonCSVTagger.cc (hunk 14 above)
Comment (emilbols) [reply RecoBTag/Combined/src/heavyIonCSVTagger.cc:91]: a better way to do the noTrack and noVertex checks would be just
`std::vector<float> tagValList = vars.getList(reco::btau::trackSip3dSig, false);`
 `bool noTrack = (tagValList.size() == 0);`
`bool noVertex = (vars.get(reco::btau::vertexCategory, -1.0) == 2)`
vertexCategory = 2 is when you have neither reco or pseudo vertex

**Comment by jpata** [comment]
@cmsbuild please test workflow 140.5611

File: RecoBTag/Combined/src/HeavyIonCSVTagger.cc [hunk 18]
```diff
@@ -80,26 +79,11 @@ float HeavyIonCSVTagger::discriminator(const TagInfoHelper &tagInfo) const {
...
-    //single value tagging var
-    else {
-      inputs[mva_var.name] = vars.get(mva_var.id, mva_var.default_value);
-      //IK: vtxMass check to check vtxType: vtxType = 2 (no vtx), vtxMass < 0, vtxType = 1 (pseudo vtx), vtxMass > 0
-      if (mva_var.name == "TagVarCSV_vertexMass") vtxMassVal = inputs[mva_var.name];
-    }
-  }
+  std::vector<float> tagValList = vars.getList(reco::btau::trackSip3dSig, false);
+  bool noTrack = (tagValList.size() == 0);
+  bool noVertex = (vars.get(reco::btau::vertexCategory, -1.0) == 2);
```
Comment (emilbols) [inline RecoBTag/Combined/src/HeavyIonCSVTagger.cc]: Sorry if i was unclear, but you still need the for loop, it fills the input vector passed to the mva. It is just these noTrack/noVertex booleans that are used to check if the jet is taggable should be outside of it.

So  
```suggestion
  bool noVertex = (vars.get(reco::btau::vertexCategory, -1.0) == 2); 
  for (auto &mva_var : variables_) {
    //vectorial tagging variable
    if (mva_var.has_index) {
      std::vector<float> vals = vars.getList(mva_var.id, false);
      inputs[mva_var.name] = (vals.size() > mva_var.index) ? vals[mva_var.index] : mva_var.default_value;   
    }
    //single value tagging var
    else {
      inputs[mva_var.name] = vars.get(mva_var.id, mva_var.default_value);
    }
  }
```

**Comment by mandrenguyen** [comment]
Tests will fail until the following PR is merged:
https://github.com/cms-data/RecoBTag-Combined/pull/36

**Comment by slava77** [comment]
test parameters:
- pull_request = cms-data/RecoBTag-Combined#36

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
Thanks for restarting the test Slava. Looks to me like the reco output is changed only for HI workflows, only for the affected tagger, as expected.

@mandrenguyen to proceed, I would like to ask:
- have you already made any larger-stats validation plots that the tagger output is as desired?
- please include a timing output to be sure

**Comment by mandrenguyen** [comment]
@jpata 
For wf 140.5611 (single thread), here's the time report:
TimeReport> Time report complete in 350.315 seconds
 Time Summary: 
 - Min event:   0.241455
 - Max event:   17.0775
 - Avg event:   1.79025
 - Total loop:  232.863
 - Total init:  114.566
 - Total job:   350.315
 - EventSetup Lock:   0
 - EventSetup Get:   0
 Event Throughput: 0.429437 ev/s


Here's the module level summary, picking out the top ones by eye
TimeReport ---------- Module Summary ---[Real sec]----
TimeReport  per event     per exec    per visit  Name
TimeReport   0.369067     0.369067     0.369067  akCs4PFJets
TimeReport   0.098232     0.098232     0.098232  puppi
TimeReport   0.082314     0.082314     0.082314  isolatedTracks
TimeReport   0.078638     0.078638     0.078638  packedPFCandidates
TimeReport   0.076910     0.076910     0.076910  secondaryVertexTagInfos
TimeReport   0.065569     0.065569     0.065569  impactParameterTagInfos
TimeReport   0.042427     0.042427     0.042427  pfCandidateToVertexAssociation
TimeReport   0.033903     0.033903     0.033903  primaryVertexAssociation
TimeReport   0.028780     0.028780     0.028780  packedCandidateTrackValidator
TimeReport   0.026883     0.026883     0.026883  patMETs
TimeReport   0.026764     0.026764     0.026764  patPFMet
TimeReport   0.023160     0.023160     0.023160  pfPileUpIsoPFBRECO
TimeReport   0.021012     0.021012     0.021012  combinedSecondaryVertexV2BJetTags
TimeReport   0.020493     0.020493     0.020493  patTrigger
TimeReport   0.018123     0.018123     0.018123  lostTracks
TimeReport   0.014771     0.014771     0.014771  hiPuRho
TimeReport   0.015700     0.015700     0.015700  ca8PFJetsCHSprunedForBoostedTaus

If it interpret correctly this module takes about 1% of the total time. 

The validation of the code and xml file where originally presented here:
https://indico.cern.ch/event/853820/contributions/3590366/attachments/1921458/3178704/HIN_BTag2018_07102019.pdf

I can of course try and check quickly that the output is still sensible after modifications made in the course of the code review, if that's what you're asking for.

**Comment by jpata** [comment]
Thanks!

Yes, combinedSecondaryVertexV2BJetTags that you modify with your computer takes about 1% according to this list, I think this is fine.

Given that some modifications were made to the code, I think making a small validation that the shapes are not changed in major ways may save time later on.

**Comment by mandrenguyen** [comment]
@jpata I don't have a setup to draw the values immediately available, and I'm having trouble getting the values out of FWLite.  
I can cout the values and they look reasonable.
For the validation of the HI miniAOD code, we plan to compare the values from our current AOD -> flat ntuple workflow to the new AOD -> miniAOD -> flat ntuple workflow, but I don't have the latter setup at the moment.

**Comment by jpata** [comment]
OK. Since I understand you (=HI) are both the main developer and consumer of this tagger, I don't see an issue to merge this now, so you can fully validatte the shape on a large scale later.

**Comment by jpata** [comment]
+reconstruction

- requires cms-data/RecoBTag-Combined#36
- introduces CSV calibrations for 2018 heavy ion data
- tests pass and show differences only in the CSV values for HI workflows, as expected
- further validation to be done by the HI group after merging

**Comment by jpata** [comment]
@mandrenguyen please add to the PR description clearly that it requires cms-data/RecoBTag-Combined#36.

**Comment by silviodonato** [comment]
merge

**Comment by santocch** [comment]
+1

**Pull request merged**